### PR TITLE
Add context-sensitive grammars

### PR DESCRIPTION
**Pull request by jaapdejong15**
Adds context-sensitive grammars.
Grammars can be defined just like normal CFGs.
Constraints can be added using `add_constraint(grammar, constraint)`.
For example, the constraint that the numbers 0-9 can only be used in an addition is defined as:
```julia
g₁ = Herb.Grammars.@csgrammar begin
    Real = |(0:9)
    Real = x
    Real = Real + Real
end
for i ∈ 1:10
    Herb.Grammars.addconstraint!(g₁, Herb.Grammars.ComesAfter(i, [12]))
end
```
Closes https://github.com/Herb-AI/Herb.jl/issues/25

### CODE REVIEW

File: src/csg/constraints.jl
Comment (sebdumancic) [inline src/csg/constraints.jl:1]: Should we immediately move the implementation of constraints to a separate `Constraints.jl` package?
That is, we can define an abstract type `Constraint` here and move all constraints and their propagators to a different package. 

Then Constraints.jl will depend on Grammar.jl, and Search.jl uses then both Grammar.jl and Constraints.jl.

This can be done later also, but just ot raise a discussion about that.

Comment (jaapdejong15) [reply src/csg/constraints.jl:1]: I think it is a very good idea to do it now, as separating might be more difficult later.

File: src/csg/constraints.jl
Comment (jaapdejong15) [reply src/csg/constraints.jl:2]: Sounds good, do you want to keep both abstract types `Constraint` and `PropagatorConstraint` so we can easily add something similar back later or should we keep only one of the two? `PropagatorConstraint` will be the only child of `Constraint` in the type tree.

File: src/csg/constraints.jl
Comment (jaapdejong15) [reply src/csg/constraints.jl]: I separated them. Therefore, there are now four related pull requests, which I all linked to this issue: https://github.com/Herb-AI/Herb.jl/issues/25

File: src/csg/csg.jl
Comment (sebdumancic) [inline src/csg/csg.jl:11]: I guess this PR came before probabilities?
So there will be no conflict, I guess.

**Review by sebdumancic — APPROVED** [review]

File: src/csg/csg.jl
Comment (jaapdejong15) [reply src/csg/csg.jl:11]: Indeed, I merged this branch in the branch for probabilistic grammars so I could continue working on that before waiting for this to be merged.

**Pull request merged**
